### PR TITLE
Un-monobehaviour compute query system

### PR DESCRIPTION
**Pull request by huwb**
**Status**: Ready to merge

This was something that I didn't get to as part of the ocean-in-edit-mode changes - making the query system standalone objects rather than monobehaviours, as there is no advantage to having them that way. Further i discovered that disabling scene reload broke the system, so i decided to move it away from behaviours now.

The resulting code in this PR is simpler, and more direct/explicit, and less stateful.

I'm not aware of any issues.

I tested in win64 editor and a standalone build.

### CODE REVIEW

**Comment by daleeidd** [comment]
I get this whenever I enter play mode:

```console
Compute shader (QueryFlow): Property (_LD_TexArray_Flow) at kernel index (0) is not set
UnityEngine.ComputeShader:Dispatch(Int32, Int32, Int32, Int32)
Crest.QueryBase:ExecuteQueries() (at Assets/Crest/Crest/Scripts/Collision/QueryBase.cs:461)
Crest.QueryBase:UpdateQueries() (at Assets/Crest/Crest/Scripts/Collision/QueryBase.cs:436)
Crest.OceanRenderer:RunUpdate() (at Assets/Crest/Crest/Scripts/OceanRenderer.cs:573)
Crest.OceanRenderer:LateUpdate() (at Assets/Crest/Crest/Scripts/OceanRenderer.cs:566)
```

Once I have entered play mode on the Whirlpool scene it goes away from then on.

All of the boats in the _boat_ scene are being thrown around for me. I have to disable the dynamic waves simulation and OceanInputs. The other scenes appear fine.

**Comment by daleeidd** [comment]
Also, it looks like this solves the slice exception I mentioned in #539. I managed to reproduce it by taking focus away from Unity and then entering play mode (about 80% reproducible).

**Comment by huwb** [comment]
Thanks @daleeidd 

I've put in the code to fix that error, and also realised that we shouldnt use the query system at all if no flow lod data is present (i.e. why run the compute shader if theres no flow data) so the error should be gone and it should be better now.

Could i please get a retest of the boat scene? It doesnt use flow (and it seems fine here) so i'm not sure what that's about. are you testing on OSX - maybe its a metal thing? Might be interesting to drop a VisualiseCollisionArea component onto a GO and see if things are behaving well.

**Comment by daleeidd** [comment]
> are you testing on OSX - maybe its a metal thing? 

It appears to be. I just tried Windows and it is fine.

If I decrease the _Time Scale_ to 0.2 it appears to be stable. Or if I change the _Fixed Timestep_ to something like 0.002 from 0.02, it appears to be stable.

> Might be interesting to drop a VisualiseCollisionArea component onto a GO and see if things are behaving well.

Looks like it is fine on Metal.

> The boat escapes the whirlpool sooner than normal.

Scratch that comment... forgot that the boat is player controlled.

**Comment by daleeidd** [comment]
The Metal issue appears tied to quantity of the floating objects. If I add ten or more boats to the _Three Boat_ scene, the issue occurs there too. But if I change the fixed time step to run more frequently (like 0.002 from 0.02) it becomes stable.

File: crest/Assets/Crest/Crest/Scripts/OceanRenderer.cs [hunk 1]
```diff
@@ -456,6 +443,12 @@ void CreateDestroyLodDatas()
                     _lodDataFlow = new LodDataMgrFlow(this);
                     _lodDatas.Add(_lodDataFlow);
                 }
+
+                if (!(FlowProvider is QueryFlow))
+                {
+                    FlowProvider.CleanUp();
```
[the remark this answers is not in the document]
Comment (huwb) [reply crest/Assets/Crest/Crest/Scripts/OceanRenderer.cs:449]: ohhh thanks

**Comment by huwb** [comment]
Do you mind doing that same test in master (not with these changes?). i feel like mattmac reported similar things in the past. I wonder whats going on..

besides that could i get your final approval on this? thanks for the help

**Comment by daleeidd** [comment]
Done.

> Do you mind doing that same test in master (not with these changes?)

I tried and it it is far more stable on master.

For the _Three Boat_ scene, I had almost 120 boats on master (compared to 20) and everything was stable. Very rarely I might see a boat dunk itself.

For the Boat scene, I could get to 40. Whilst here, just the scene without any changes and they go wild. I'll get a video on discord.

**Comment by daleeidd** [comment]
Cannot reproduce with OS video capture running. I've heard that before :\

**Pull request merged**